### PR TITLE
[FSDP2] Added `set_reshard_after_backward`

### PR DESCRIPTION
**Pull request by awgu**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #124513
* #120256
* __->__ #124319

This PR adds a `set_reshard_after_backward` method to allow disabling resharding after backward. `reshard_after_backward=False` can be used with `reshard_after_forward=False` to implement "ZeRO-1", where there is only all-gather on the first microbatch forward and reduce-scatter on the last microbatch backward.

```
for microbatch_idx, microbatch in dataloader:
    is_last_microbatch = microbatch_idx == num_microbatches - 1
    model.set_requires_gradient_sync(is_last_microbatch)
    model.set_reshard_after_backward(is_last_microbatch)
    model.set_is_last_backward(is_last_microbatch)
    microbatch_fwd_bwd(model, microbatch, microbatch_idx)
```

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k

### CODE REVIEW

File: torch/distributed/_composable/fsdp/_fsdp_param_group.py [hunk 1]
```diff
@@ -309,7 +312,8 @@ def post_backward(self, *unused: Any):
         self._training_state = TrainingState.POST_BACKWARD
         with torch.profiler.record_function("FSDP::post_backward_reshard"):
             if not self.reduce_grads:
-                self.reshard()
+                if self.reshard_after_backward:
```
Comment (awgu) [inline torch/distributed/_composable/fsdp/_fsdp_param_group.py:315]: This is the main expected use case: do not reduce gradients, and do not reshard parameters.

File: torch/distributed/_composable/fsdp/_fsdp_param_group.py [hunk 2]
```diff
@@ -320,7 +324,8 @@ def post_backward(self, *unused: Any):
                     fsdp_params_with_grad.append(fsdp_param)
                     unsharded_grads.append(fsdp_param.unsharded_grad_data)
                     fsdp_param.unsharded_param.grad = None
-            self.reshard()
+            if self.reshard_after_backward:
```
Comment (awgu) [inline torch/distributed/_composable/fsdp/_fsdp_param_group.py:327]: However, it is also easy for us to support the other case: _do_ reduce gradients, but do not reshard parameters.

If people want to play around with different collective schedules during gradient accumulation, then this provides an extra lever.

File: torch/distributed/_composable/fsdp/_fsdp_param_group.py (hunk 1 above)
Comment (weifengpy) [reply torch/distributed/_composable/fsdp/_fsdp_param_group.py:315]: is the use case mainly for pipeline parallel?

**Review by weifengpy — APPROVED** [review]

File: torch/distributed/_composable/fsdp/_fsdp_param_group.py (hunk 1 above)
Comment (awgu) [reply torch/distributed/_composable/fsdp/_fsdp_param_group.py:315]: This is useful for any kind of gradient accumulation, where pipeline parallelism is one example.

**Comment by awgu** [comment]
```
2024-04-19T15:16:34.7142216Z # Prune all of the docker images
2024-04-19T15:16:34.7142516Z docker system prune -af
2024-04-19T15:16:34.7150054Z shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
2024-04-19T15:16:34.7150375Z ##[endgroup]
2024-04-19T15:16:34.7309627Z Cannot connect to the Docker daemon at unix:///run/docker/docker.sock. Is the docker daemon running?
2024-04-19T15:16:34.7467484Z "docker stop" requires at least 1 argument.
2024-04-19T15:16:34.7468100Z See 'docker stop --help'.
2024-04-19T15:16:34.7468277Z 
2024-04-19T15:16:34.7468438Z Usage:  docker stop [OPTIONS] CONTAINER [CONTAINER...]
2024-04-19T15:16:34.7468697Z 
2024-04-19T15:16:34.7468819Z Stop one or more running containers
2024-04-19T15:16:34.7604790Z "container prune" requires API version 1.25, but the Docker daemon API version is 1.24
2024-04-19T15:16:34.7628585Z ##[error]Process completed with exit code 1.
```
[pull / linux-jammy-py3.8-gcc11-pch / build](https://github.com/pytorch/pytorch/actions/runs/8755724518/job/24030506185) failure looks unrelated
